### DISCUSSION TITLE
An improved applicability syntax

### DISCUSSION THREAD

**Discussion by Moult** [buildingSMART/IDS#324] [Ideas]
In IfcOpenShell, we've designed a query syntax that is inspired by IDS but seems to overcome most of the common shortcomings people find with IDS applicability facets.

I've been receiving lots of feedback saying "Why doesn't IDS use this", so I thought I'd post it here: https://docs.ifcopenshell.org/ifcopenshell-python/selector_syntax.html#filtering-elements

**Comment by NickNisbet** [comment]
Yes, but we should not bury Selections and Exceptions in complex syntax.
In my suggestion they are explicit, and can be recursive if necessary.

Facets should not allow the applicability, selection or exception (or even a requirement) by guid. This is a major flaw in  model checking applications - it allows people to cheat/delude/mislead by faking the filter ('classification') rule.

Similarly the facets should not depend on global or local parameters. Really useful for reporting though. Missing 'system' or 'zone' or workpackage' or more generally 'group'

**Reply by Moult** [reply]
Sure, you don't need to follow the syntax, you can still use XML if you want - but the princples can still apply.

GUID is also not required of course.

However this approach allows you to define NOT, OR, and UNION selections.

**Comment by NickNisbet** [comment]
Covering AND  OR NOT is exactly what my four section approach does, in a meaningful, manageable and structured way. UNION means OR. These four concept are more natural, rather than mathematical.

Result = not_applicable or not_selected or excepted or as_required
      not_applicable = not(and( a1, a2... an)
      not_selected = not(or(s1,s2,...sn)
      excepted = or(e1,e2,...en)
      as_required=and(r1,r2,...rn)
